### PR TITLE
gihan_cmp

### PR DESCRIPTION
**Pull request by gihanpanapitiya**
Adding a workflow to find prediction errors of feature domains. Can be used to compare models belonging to different architectures. 

Files modified to add this workflow:

        modified:   workflows/cmp-cv/swift/workflow.sh
        modified:   workflows/cmp-cv/swift/workflow.swift
        modified:   workflows/cmp-cv/test/models-1.txt
        modified:   workflows/cmp-cv/test/test-small-1.sh
        modified:   workflows/cmp-cv/test/upf-1.txt
        new file:   workflows/common/python/compare.py
        modified:   workflows/common/python/model_runner.py
        modified:   workflows/common/sh/model.sh
        modified:   workflows/common/swift/obj_app.swift

### CODE REVIEW

File: workflows/cmp-cv/swift/workflow.sh [hunk 1]
```diff
@@ -17,7 +17,7 @@ usage()
   echo "CMP-CV: usage: workflow.sh SITE EXPID CFG_SYS PLAN"
 }
 
-if (( ${#} != 4 ))
```
Comment (rajeeja) [inline workflows/cmp-cv/swift/workflow.sh:20]: This will break other workflows that supply 4 arguments.

File: README.adoc [hunk 2]
```diff
@@ -1 +1,37 @@
...
+```
+{"id": "RUN000", "epochs": 1, "model_name": "DrugCell", "candle_image": "/path/to/sif/DrugCell.sif"}
+{"id": "RUN001", "epochs": 2, "model_name": "DrugCell", "candle_image": "/path/to/sif/DrugCell.sif"}
+{"id": "RUN002", "epochs": 1, "model_name": "SWnet_CCLE", "candle_image": "/path/to/sif/SWnet.sif"}
+{"id": "RUN003", "epochs": 2, "model_name": "SWnet_CCLE", "candle_image": "/path/to/sif/SWnet.sif"}
+```
+
+### Running the specific example at workflows/cmp-cv/test
+
+- Clone Supervisor https://github.com/gihanpanapitiya/Supervisor/tree/gihan_cmp
```
Comment (rajeeja) [inline README.adoc]: Put generic instructions, not pointing to your specific repo.

File: README.adoc [hunk 3]
```diff
@@ -1 +1,37 @@
...
+- Clone the DrugCell and SWnet model directories from https://github.com/gihanpanapitiya/DrugCell/tree/to_candle and https://github.com/gihanpanapitiya/SWnet/tree/to_candle
```
Comment (rajeeja) [inline README.adoc:21]: This is fine, add a line about how to get the .sif files.

File: workflows/common/python/compare.py [hunk 4]
```diff
@@ -0,0 +1,162 @@
+import os
+import pandas as pd
+import pandas as pd
+import numpy as np
+from sklearn.metrics import mean_squared_error
+
+conditions = pd.DataFrame([[ 'nAromAtom' , 5, 10 ],
```
Comment (rajeeja) [inline workflows/common/python/compare.py:19]: Please add some context on what are we comparing and text on cancer drug discovery domain for this python file.

File: workflows/common/python/model_runner.py [hunk 5]
```diff
@@ -171,50 +171,62 @@ def run(hyper_parameter_map, obj_return):
...
+            if history is not None:
+                if history == "EPOCHS_COMPLETED_ALREADY":
+                    result, history_result = "EPOCHS_COMPLETED_ALREADY", None
+                else:
+                    result, history_result = get_results(history, obj_return,
+                                                        epochs)
+        else:
+            result, history_result = "RUN_EXCEPTION", None
+
+    elif framework == 'pytorch':
```
Comment (rajeeja) [inline workflows/common/python/model_runner.py:222]: Good changes here, will this be true for any pytorch model?

It'd be good to launch our Jenkins testing on this branch before merging.

File: workflows/common/sh/model.sh [hunk 6]
```diff
@@ -44,7 +47,7 @@ RUNID=$4
 if [[ $CANDLE_MODEL_TYPE = "SINGULARITY" ]]
 then
   # TODO: Rename "instance" to "run"
-  INSTANCE_DIRECTORY=$CANDLE_DATA_DIR/$MODEL_NAME/Output/$EXPID/$RUNID
+  INSTANCE_DIRECTORY=$CANDLE_DATA_DIR/$MODEL_NAME/Output/$EXPID/$RUNID # gihan -removing $ in MODEL_NAME
```
Comment (rajeeja) [inline workflows/common/sh/model.sh]: remove the comment.

File: workflows/common/swift/obj_app.swift [hunk 7]
```diff
@@ -9,11 +9,13 @@
 */
 (string obj_result) obj(string params,
                         string expid,
-                        string runid)
+                        string runid,
+                        string model_name,
+                        string candle_image)
```
Comment (rajeeja) [inline workflows/common/swift/obj_app.swift]: Again, this function signature change would break other workflows @j-woz what do you think is the best work around? 

How about another obj for singularity say '`obj_singularity`'?

**Review by rajeeja — CHANGES_REQUESTED** [review]

File: workflows/cmp-cv/swift/workflow.sh (hunk 1 above)
Comment (gihanpanapitiya) [reply workflows/cmp-cv/swift/workflow.sh:20]: Since this workflow is only applicable for cmp-cv, adding this workflow.sh will not affect the other workflows, right?

File: README.adoc (hunk 3 above)
Comment (gihanpanapitiya) [reply README.adoc:21]: Added the command to get the .sif files

File: workflows/common/python/model_runner.py (hunk 5 above)
Comment (gihanpanapitiya) [reply workflows/common/python/model_runner.py:222]: This is not pytorch specific. We could also have this for keras. This change assumes that after training the model, the run() method returns prediction scores like loss, rmse for validation and test sets. We can have this for all the models, but currently not every drug response model returns these values.

File: workflows/common/swift/obj_app.swift (hunk 7 above)
Comment (gihanpanapitiya) [reply workflows/common/swift/obj_app.swift]: If we could export the MODEL_NAME and CANDLE_IMAGE in workflow.swift, then we wouldn't need to modify obj_app.swift.

I couldn't find a way to set env variables in .swift files. If we could, we can replace the following lines in cmp-cv/swift/workflow.swift,
  model_name = json_get(gparam, "model_name");
  candle_image = json_get(gparam, "candle_image");
  
  with something like,
  
  export MODEL_NAME=json_get(gparam, "model_name");
  export CANDLE_IMAGE = json_get(gparam, "candle_image");

File: workflows/cmp-cv/swift/workflow.sh (hunk 1 above)
Comment (rajeeja) [reply workflows/cmp-cv/swift/workflow.sh:20]: True that!

File: workflows/common/sh/model.sh (hunk 6 above)
Comment (gihanpanapitiya) [reply workflows/common/sh/model.sh]: Done.

File: README.adoc (hunk 2 above)
Comment (gihanpanapitiya) [reply README.adoc]: Done.

File: workflows/common/sh/model.sh [hunk 8]
```diff
@@ -24,7 +24,7 @@ usage()
   echo "If SH_TIMEOUT is set, we run under the shell command timeout"
 }
 
-if (( ${#} != 4 ))
+if (( ${#} != 6 ))
```
Comment (rajeeja) [inline workflows/common/sh/model.sh]: This breaks the other models, so just say something like:

if num args is 4 then:

load first 4

then if it is 6:

load the other 2.

else:
throw and error.

File: workflows/common/swift/obj_app.swift (hunk 7 above)
Comment (rajeeja) [reply workflows/common/swift/obj_app.swift]: Let's call in `obj_container` we need to pass them to run multiple models with our function:
run_model(model_sh, params, expid, runid, model_name, candle_image)

Comment (gihanpanapitiya) [reply workflows/common/swift/obj_app.swift]: Now there is a new obj, obj_container added to the workflows/common/swift. obj_app.swift is not changed.

File: workflows/common/python/compare.py (hunk 4 above)
Comment (rajeeja) [reply workflows/common/python/compare.py:19]: Add a top level header stating this is for cancer drug problem and can be tailored for specific applications

Comment (gihanpanapitiya) [reply workflows/common/python/compare.py:19]: Added a description about the code and its usefulness.

**Review by rajeeja — APPROVED** [review]
